hearth/lib/hearth/event_stream/encoder.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+module Hearth
+  # Module for EventStreams.
+  module EventStream
+    # Event Stream encoder. Replaces the request body on operations with input
+    # event streams and provides the ability to signal, encode,
+    # sign (with state) and send event messages.
+    class Encoder
+      # @param [Object] message_encoder a protocol specific message encoder.
+      # @param [IO | Message] the initial event body. When the initial body
+      #   is a message, it is treated as a special initial-request event and
+      #   the read method is defined on this encoder instance, allowing the
+      #   http2 client to send it.
+      def initialize(message_encoder:, initial_event_body:)
+        @message_encoder = message_encoder
+        @initial_event_body = initial_event_body
+
+        @prior_signature = nil
+        @sign_event = nil
+
+        return unless @initial_event_body.is_a?(Message)
+
+        define_singleton_method(:read) do
```

Comment:
What's with this? What about a class method? Why must it be singleton?

Reply:
This is to handle initial requests with bodies. The http2 client checks for respond_to(:read) on the body of the request, and if that method exists, it will read the data from it and send.

I've refactored this and created a new class `EncoderWithInitialMessage` that extends `Encoder` and adds the `read` method.  The `EventStream::Handlers` now does the check if the request body is a Message, and if so, creates the `EncoderWithInitialMessage` instead.